File: jm/diag/kwp2000.cpp
#include "kwp2000.h"
#include "../utils.h"

namespace JM
{
namespace Diag
{

KWP2000::KWP2000()
{

}

KWP2000::~KWP2000()
{

}

std::size_t KWP2000::pack(const boost::uint8_t* src,
                          std::size_t srcLength,
                          boost::uint8_t *tar,
                          std::size_t tarLength,
                          boost::system::error_code &ec)
{
    std::size_t pos = 0;
    boost::uint8_t checksum = 0;
    std::size_t i = 0;

    if (_mode == Mode8X)
    {
        tar[pos++] = low_byte(0x80 | srcLength);
        tar[pos++] = low_byte(_opts.targetAddress);
        tar[pos++] = low_byte(_opts.sourceAddress);
        memcpy(tar + pos, src, srcLength);
        pos += srcLength;
    }
    else if (_mode == ModeCX)
    {
        tar[pos++] = low_byte(0xC0 | srcLength);
        tar[pos++] = low_byte(_opts.targetAddress);
        tar[pos++] = low_byte(_opts.sourceAddress);
        memcpy(tar + pos, src, srcLength);
        pos += srcLength;
    }
    else if (_mode == Mode80)
    {
        tar[pos++] = low_byte(0x80);
        tar[pos++] = low_byte(_opts.targetAddress);
        tar[pos++] = low_byte(_opts.sourceAddress);
        tar[pos++] = low_byte(srcLength);
        memcpy(tar + pos, src, srcLength);
        pos += srcLength;
    }
    else if (_mode == Mode00)
    {
        tar[pos++] = 0x00;
        tar[pos++] = low_byte(srcLength);
        memcpy(tar + pos, src, srcLength);
        pos += srcLength;
    }
    else if (_mode == ModeXX)
    {
        tar[pos++] = low_byte(srcLength);
        memcpy(tar + pos, src, srcLength);
        pos += srcLength;
    }
    else
    {
        ec = boost::system::error_code(boost::system::errc::protocol_error,
                                       boost::system::get_system_category());
    }

    for (i = 0; i < pos; i++)
    {
        checksum += tar[i];
    }
    tar[pos++] = checksum;
    return pos;
}

std::size_t KWP2000::unpack(const boost::uint8_t *src,
                            std::size_t srcLength,
                            boost::uint8_t *tar,
                            std::size_t tarLength,
                            boost::system::error_code &ec)
{
    std::size_t length = 0;

    if (src[0] > 0x80)
    {
        length = src[0] - 0x80;
        if (src[1] != _opts.sourceAddress)
        {
            ec = boost::system::error_code(boost::system::errc::bad_message,
                                           boost::system::get_system_category());
            return 0;
        }
        if (length != (srcLength - KWP8X_HEADER_LENGTH - KWP_CHECKSUM_LENGTH))
        {
            length = src[0] - 0xC0; /* For KWPCX */
            if (length != (srcLength - KWPCX_HEADER_LENGTH - KWP_CHECKSUM_LENGTH))
            {
                ec = boost::system::error_code(boost::system::errc::message_size,
                                               boost::system::get_system_category());
                return 0;
            }
        }
        memcpy(tar, src + KWP8X_HEADER_LENGTH, length);
    }
    else if (src[0] == 0x80)
    {
        length = src[0];
        if (src[1] != _opts.sourceAddress)
        {
            ec = boost::system::error_code(boost::system::errc::bad_message,
                                           boost::system::get_system_category());
            return 0;
        }

        if (length != (srcLength - KWP80_HEADER_LENGTH - KWP_CHECKSUM_LENGTH))
        {
            ec = boost::system::error_code(boost::system::errc::message_size,
                                           boost::system::get_system_category());
            return 0;
        }
        memcpy(tar, src + KWP80_HEADER_LENGTH, length);
    }
    else if (src[0] == 0x00)
    {
        length = src[1];
        if (length != (srcLength - KWP00_HEADER_LENGTH - KWP_CHECKSUM_LENGTH))
        {
            ec = boost::system::error_code(boost::system::errc::message_size,
                                           boost::system::get_system_category());
            return 0;
        }
        memcpy(tar, src + KWP00_HEADER_LENGTH, length);
    }
    else
    {
        length = src[0];
        if (length != (srcLength - KWP00_HEADER_LENGTH - KWP_CHECKSUM_LENGTH))
        {
            ec = boost::system::error_code(boost::system::errc::message_size,
                                           boost::system::get_system_category());
            return 0;
        }
        memcpy(tar, src + KWP00_HEADER_LENGTH, length);
    }
    return length;
}

boost::system::error_code
KWP2000::config(const KWP2000::Options &opts, boost::system::error_code &ec)
{
    memcpy(&_opts, &opts, sizeof (Options));
    ec = boost::system::error_code();
    return ec;
}

}
}

File: JM/Diag/KWP2000.cs
using System;
using JM.Core;

namespace JM.Diag
{
    public class KWP2000 : IProtocol
    {
        protected KWPOptions options;
        private KWPMode mode;

        public const int KWP8X_HEADER_LENGTH = 3;
        public const int KWPCX_HEADER_LENGTH = 3;
        public const int KWP80_HEADER_LENGTH = 4;
        public const int KWPXX_HEADER_LENGTH = 1;
        public const int KWP00_HEADER_LENGTH = 2;
        public const int KWP_CHECKSUM_LENGTH = 1;
        public const int KWP_MAX_DATA_LENGTH = 128;

        public KWP2000()
        {
            mode = KWPMode.Mode8X;
        }

        public byte[] Pack(byte[] data, int offset, int count)
        {
            int pos = 0;
            byte checksum = 0;
            int i = 0;
            byte[] result = null;

            if (mode == KWPMode.Mode8X)
            {
                result = new byte[KWP8X_HEADER_LENGTH + count + KWP_CHECKSUM_LENGTH];
                result[pos++] = Utils.LowByte(0x80 | count);
                result[pos++] = Utils.LowByte(options.TargetAddress);
                result[pos++] = Utils.LowByte(options.SourceAddress);
            }
            else if (mode == KWPMode.ModeCX)
            {
                result = new byte[KWPCX_HEADER_LENGTH + count + KWP_CHECKSUM_LENGTH];
                result[pos++] = Utils.LowByte(0xC0 | count);
                result[pos++] = Utils.LowByte(options.TargetAddress);
                result[pos++] = Utils.LowByte(options.SourceAddress);
            }
            else if (mode == KWPMode.Mode80)
            {
                result = new byte[KWP80_HEADER_LENGTH + count + KWP_CHECKSUM_LENGTH];
                result[pos++] = Utils.LowByte(0x80);
                result[pos++] = Utils.LowByte(options.TargetAddress);
                result[pos++] = Utils.LowByte(options.SourceAddress);
                result[pos++] = Utils.LowByte(count);
            }
            else if (mode == KWPMode.Mode00)
            {
                result = new byte[KWP00_HEADER_LENGTH + count + KWP_CHECKSUM_LENGTH];
                result[pos++] = Utils.LowByte(0x00);
                result[pos++] = Utils.LowByte(count);
            }
            else if (mode == KWPMode.ModeXX)
            {
                result = new byte[KWPXX_HEADER_LENGTH + count + KWP_CHECKSUM_LENGTH];
                result[pos++] = Utils.LowByte(count);
            }
            else
            {
                throw new ArgumentException();
            }

            Array.Copy(data, offset, result, pos, count);
            pos += count;

            for (i = 0; i < pos; i++)
            {
                checksum += result[i];
            }
            result[i] = checksum;
            return result;
        }

        public byte[] Unpack(byte[] data, int offset, int count)
        {
            int length = 0;
            byte[] result = null;

            if ((data[offset] & 0xFF) > 0x80)
            {
                length = (data[offset] & 0xFF) - 0x80;
                if (data[offset + 1] != options.SourceAddress)
                {
                    return null;
                }
                if (length != (count - KWP8X_HEADER_LENGTH - KWP_CHECKSUM_LENGTH))
                {
                    length = data[offset] - 0xC0; // For KWPCX 
                    if (length != (count - KWPCX_HEADER_LENGTH - KWP_CHECKSUM_LENGTH))
                    {
                        return null;
                    }
                }
            }
            else if ((data[offset] & 0xFF) == 0x80)
            {
                length = data[0] & 0xFF;
                if (data[offset + 1] != options.SourceAddress)
                {
                    return null;
                }

                if (length != (count - KWP80_HEADER_LENGTH - KWP_CHECKSUM_LENGTH))
                {
                    return null;
                }
            }
            else if (data[offset] == 0x00)
            {
                length = data[offset + 1] & 0xFF;
                if (length != (count - KWP00_HEADER_LENGTH - KWP_CHECKSUM_LENGTH))
                {
                    return null;
                }
            }
            else
            {
                length = data[offset] & 0xFF;
                if (length != (count - KWP00_HEADER_LENGTH - KWP_CHECKSUM_LENGTH))
                {
                    return null;
                }
            }

            result = new byte[length];
            Array.Copy(data, offset + KWP8X_HEADER_LENGTH, result, 0, length);
            return result;
        }

        public void Config(KWPOptions opts)
        {
            options.LinkMode = opts.LinkMode;
            options.LLine = opts.LLine;
            options.MsgMode = opts.MsgMode;
            options.SourceAddress = opts.SourceAddress;
            options.TargetAddress = opts.TargetAddress;
        }
    }
}